storage/src/main/scala/coop/rchain/storage/regex/Fsm.scala

```diff
@@ -0,0 +1,806 @@
+        val transition = fsm.transitions.get(fsmState)
+        if (transition.isDefined) {
+          val nextState = transition.get.get(currentSymbol)
+          if (nextState.isDefined) {
+            val changes = connectAll(fsmIndex, nextState.get)
+            nextStates ++= changes
+          }
+        }
+      }
+
+      if (nextStates.isEmpty) {
+        None
+      } else {
+        Some(nextStates)
+      }
+    }
+
+    crawl(alphabet, initalStates, isFinal, follow).Reduced
+  }
+}
+
+/**
+  * A Finite State Machine or FSM has an alphabet and a set of states. At any
+  * given moment, the FSM is in one state. When passed a symbol from the
+  * alphabet, the FSM jumps to another state (or possibly the same state).
+  * The 'transtions' map indicates where to jump.
+  * One state is nominated as a starting state. Zero or more states are
+  * nominated as final states. If, after consuming a string of symbols,
+  * the FSM is in a final state, then it is said to "accept" the string.
+  * This class also has some pretty powerful methods which allow FSMs to
+  * be concatenated, alternated between, multiplied, looped (Kleene star
+  * closure), intersected, and simplified.
+  * The majority of these methods are available using operator overloads.
+  */
+case class Fsm(alphabet: Set[Char],
+               states: Set[Int],
+               initialState: Int,
+               finalStates: Set[Int],
+               transitions: Map[Int, Map[Char, Int]]) {
+  require(alphabet != null, "alphabet can't be null")
+  require(states != null, "alphabet can't be null")
+  require(finalStates != null, "finals can't be null")
+  require(transitions != null, "map can't be null")
+
+  require(states.contains(initialState), "Initial state " + initialState + " must be one of states")
+  require(finalStates.forall(fin => states.contains(fin)), "Final states must be a subset of states")
+
+  for((mapEntry, stateTransitionEntry) <- transitions;
+      (nextSymbol, nextState) <- stateTransitionEntry) {
+    require(states.contains(mapEntry), "Map state " + mapEntry + " must be one of states")
+    require(alphabet.contains(nextSymbol), "Transition symbol " + nextSymbol + " -> " + nextState + " must be one of alphabet")
+    require(states.contains(nextState), "Transition state  " + nextSymbol + " -> " + nextState + " must be one of states")
+  }
+
+  /**
+    * True if FSM alphabet contains Fsm.AnythingElse and thus can accept unknown characters
+    */
+  val HasAnythingElse: Boolean = alphabet.contains(Fsm.AnythingElse)
+
+  /**
+    * A state is "live" if a final state can be reached from it.
+    */
+  def IsLive(state: Int): Boolean = {
+    val reachable = Queue(state)
+    val checked = HashSet(state) //regexes usually have quite small number of states, sorted set should be faster in most cases
+
+    while (reachable.nonEmpty) {
+      val reachableState = reachable.dequeue()
+
+      if (finalStates.contains(reachableState)) {
+        return true
```

Comment:
Original implementation throws exception here. However exceptions for control flow is the worst use of exceptions ever. I tried to rewrite that loop with functional flow like "....exists(...)", but Queue iterator is derived from LinkedList iterator and doesn't iterate over newly added items. I can write my own implicit iterator, but that's look like overkill for so simple part. Finally I only left with two options - 1) introduce a mutable variable, test it in the loop head, and modify in the loop body; or 2) use return. I decided to use "return", since it looks more readable (imho), than additional control variable. Please let me know if mutable control variable is preferred even in such cases.